content/clause-interleaving.md

```diff
@@ -0,0 +1,178 @@
+* Methods containing curried type clauses like `def foo[A][B](x: B)` cannot be represented in this way, as polymorphic methods always have to have a term parameter right after.
+* It is inefficient, as many closures must be created for each call to `getOrElse` (one per term clause to the right of the first non-initial type clause).
+* Same problem as above with overloading
+
+## Proposed solution
+### High-level overview
+
+To solve the above problems, we propose to generalize method signatures so that they can have multiple type parameter lists, interleaved with term parameter lists and using parameter lists.
+
+For the heterogeneous key-value store example, this allows to define `getOrElse` as follows:
+~~~ scala
+def getOrElse(key: Key)[V >: key.Value](default: => V): V = …
+~~~
+It provides the best of all worlds:
+* A convenient API at call site
+* A single point of documentation
+* Efficiency, since the method erases to a single JVM method with signature `getOrElse(Object,Object)Object`
+
+### Specification
+We amend the syntax of def parameter clauses as follows:
+
+~~~
+DefDcl                 ::=  DefSig ‘:’ Type
+DefDef                 ::=  DefSig [‘:’ Type] ‘=’ Expr
+DefSig                 ::=  id [DefParamClauses] [DefImplicitClause]
+DefParamClauses        ::=  DefParamClauseChunk {DefParamClauseChunk}
+DefParamClauseChunk    ::=  [DefTypeParamClause] TermOrUsingParamClause {TermOrUsingParamClause}
+TermOrUsingParamClause ::=  DefTermParamClause
+                         |  UsingParamClause
+DefTypeParamClause     ::=  [nl] ‘[’ DefTypeParam {‘,’ DefTypeParam} ‘]’
+DefTypeParam           ::=  {Annotation} id [HkTypeParamClause] TypeParamBounds
+DefTermParamClause     ::=  [nl] ‘(’ [DefTermParams] ‘)’
+UsingParamClause       ::=  [nl] ‘(’ ‘using’ (DefTermParams | FunArgTypes) ‘)’
+DefImplicitClause      ::=  [nl] ‘(’ ‘implicit’ DefTermParams ‘)’
+DefTermParams          ::=  DefTermParam {‘,’ DefTermParam}
+DefTermParam           ::=  {Annotation} [‘inline’] Param
+Param                  ::=  id ‘:’ ParamType [‘=’ Expr]
+~~~
+
+The main rules of interest are `DefParamClauses` and `DefParamClauseChunk`, which now allow any number of type parameter clauses, term parameter clauses and using parameter clauses, in any order as long as there are no two adjacent type clauses.
+
+Note that these are also used for the right-hand side of extension methods, clause interleaving thus also applies to them.
+
+It is worth pointing out that there can still only be at most one implicit parameter clause, which, if present, must be at the end.
+
+The type system and semantics naturally generalize to these new method signatures.
+
+### Restrictions
+
+#### Type Currying
+Type parameters cannot be curried (having two type clauses next to each other), as this would allow partial type inference, there is a big concern it would become a recommended norm to _always_ curry type parameters.
```

Comment:
> I guess the worry is that over-usage of type currying would make the code unnecessarily cryptic and more difficult to read for humans.
```suggestion
Currying type clauses enables partial type inference, as the left clause can be specified while the right one is not.
As this is a very useful feature, we expect people would use it liberally, and recommending the curried form.
We are uncertain about the readability of the resulting methods, we have therefore decided to not include type currying as part of this proposal.
```

Reply:
@bjornregnell What do you think of the above ?

Reply:
Looks very Good To Me!